### DISCUSSION TITLE
Release process retrospective for 3.7.0

### DISCUSSION THREAD

**Discussion by hoffie** [jamulussoftware/jamulus#1302] [General]
With 3.7.0 released, I think we should check if there are things to improve, especially considering that this was our first release by this team.

This is mainly intended for the @jamulussoftware/maindevelopers and @jamulussoftware/translators, but any contributor or watcher should feel free to chime in. :)

Some things I can think of:

**The good**
- We have a release with exciting improvements and people using it already! :)
- We tried documenting all the required steps, especially in the [3.7.0 release checklist](https://github.com/jamulussoftware/jamulus/issues/1086).

**The bad**
- The release took longer than expected in all stages:
  - It took some time until we had decided that we should start a release process.
  - It took some time to figure out details such as translation handling.
  - The actual release took time due to bugs in last-minute website PRs, due to lack of access to SourceForge (sorted out).
- While the checklist was useful, it got too long.
  - The checklist was overwritten multiple times (I think mostly myself!?) causing items to get lost and having to restore them from the edit history.
  - At least translations should rather be done in their own per-language-per-release issues so that they can be tracked properly. This might make sense for other more complicated or long-running tasks as well.

Now that we know about such shortcomings, it will be possible to work around them. I also think that the learning process and starting to create some automations took some time for the current release, but may be quicker in the upcoming ones.

I'm planning to aggregate everything I can think of here: https://github.com/jamulussoftware/jamuluswebsite/issues/340

All in all, I think it worked really well, but there are always things to improve on. :)

**Comment by gilgongo** [comment]
Many thanks for this. 

The release checklist was probably a bit complicated because we needed to discuss and modify it as we went along. Maybe future releases won't need to be quite as granular (or will be more automated). Don't forget also that the Tracking Project was only really functional about half way through. I found it useful after that, but I don't know if anyone else was looking at it,

As to process itself: we have discussed the idea of creating a "rhythm" of releases based on a fixed time after the previous release (and if this gets late for some days, so be it). This might help a number of things but mainly because translation (while it's a very good thing!) is also a significant factor in our flexibility. Unlike engineers, translators can't  postpone to the next release. And the amount of translation per release is also hard to predict (although we can control it to some extent).

Such a "target date" system might also help clarify our status overall and help with decisions to "de-tag" if it looks like we'll be giving the translators too much to do or we feel we need a beta release first, etc. Personally, I think we should be a bit cautious about the frequency at first and aim for "release plus 8 week" cycle because we need to give translators at least 2 weeks, based on previous history, and preferably maybe 3. We could also say that beta releases don't depend on translation, so betas would be outside the "target date" cycle and could happen any time (we've not talked much about testing I don't think too ...).

EDIT: This may also relate to my frustratingly un-formed thoughts about a road map. We have "triage" stage in Tracking but no high-level criteria for what should be worked on really. A road map might at least help clarify what we think is worth putting effort into now rather than later. But does that imply a "Jedi Council"? Perhaps that's not very FOSS...

**Comment by pljones** [comment]
When I first got involved with Jamulus, translations came along whenever the translator provided them - frequently long after the English and German from Volker.  I think it's a really good improvement - and should get called out as such here - that we aimed to keep the entire release updated across all supported languages.

As I mentioned during the process, I'd like to see the translators getting each issue and doing the translations before that issue is closed.  The code (or primary documentation, even) changes may already have been merged at this point.  So it would mean _always_ having a separate issue that was the "true" tracking point on the plan, with the PR just for the code review -- unless, of course, there was no change to UI display.  As an alternative (lighter weight and now I've thought, probably preferable), we may choose to say that unless the translations, where needed, are on the PR, the PR won't be merged.  I don't feel comfortable with "dumping" the release onto the translators and saying "OK, sort that out".  Being part of the development of a feature seems more inclusive.  But I can't speak for the translators, of course!

----

Cycles, rhythm, etc...  Not something I'm keen on.  KISS should apply.  Don't do too much at once.  If that means fix immediate bugs plus get the "most ready" feature delivered and nothing else (but including all the docs and translations) every couple of weeks, fine.  If we decide that's too often, it's easier to slow down than speed up - just tag a beta and move on.

----

Tools...  I used both the checklist and Tracking Project.  Finding them was the most frustrating part.  Maybe have another repo -- "jamulussoftware/pm" -- where the `README.md` has "Useful links for project management" and little else... :)

Compared with what I'm used to at work, both PRs and Issues have far more _discussion_ on them.  That all takes place in chat (Slack) or video calls where needed -- but it's a small team where we know who we're working with.  I'm just not really comfortable that the tools on Github are exactly designed for being used in the way they're being used.

I know we're "closing and moving to discussion" when something is clearly starting to look precarious...  Maybe _everything_ should be required to start as a discussion, then a nominated person (by the people in the discussion, agreed by themselves) raises the Issue as those in the discussion have already agreed.

I'd rather not have to review _every_ discussion -- I don't have time -- but I do want to see the Issues, to check whether I think they'll have significant impact one way or another.  Yes, it may end up back in the discussion getting the issue closed (but at least I'll know that one needs following).

**Reply by gilgongo** [reply]
> unless the translations ... are on the PR, the PR won't be merged.

Not sure I understand - do you mean creating a PR for each language in advance (and not an issue as we first discussed)? Agree we should maybe make more of the fact that we fully translate each release BTW. 

The issue of cycles is a tricky one. You say it's not something you're keen on, but then describe a two-week cycle :-) I would say the advantage of naming a target date right after a release is that people can use that date to plan their time. Fine if things slip, but at least people can put something in the diary, as it were.

> Maybe have another repo -- "jamulussoftware/pm" 

Fear not, there will be a page on the website for all this shortly.

> both PRs and Issues have far more discussion on them. 

Indeed, and in fact I have gently been enforcing the principle that unless it's a reasonably clear bug report, everything should be required to start as a discussion (don't forget that Discussions are a new GH feature, so GH regulars won't immediately know to use them and not Issues). In my draft guidelines (which I guess we could have as a sticky?) I [explain the deal](https://docs.google.com/document/d/1_h6111X3I8g5UvkrtwHrx7263U_W0tbu-Rnj0UmzmAs/edit). But it'll take a while for people to get the message.  

The bottom line is that the Issues should be an agreed list of _actionable_ things. In fact they are now pretty much, just need to keep them that way. Whether we can get people into the idea of collaboratively agreeing specs for work to put on the list after discussion remains to be seen though - I've tried encouraging people to do this but they are naturally reluctant if they think it means they "own" the issue perhaps. 

BTW related to your statement about lots of discussions on PRs and Issues, I have also thought we should edit the contents of Issues so that the "ticket" reflects changes made in the (inevitable) comments on it. I do that for my own tickets, but for some reason others seem happy to have good information scattered around the comments. Maybe that's just me though.

> I'm just not really comfortable that the tools on Github are exactly designed for being used in the way they're being used.

I'd reserve judgement on that until we've done another couple of releases. The system is still bedding down.

**Reply by softins** [reply]
> As I mentioned during the process, I'd like to see the translators getting each issue and doing the translations before that issue is closed. The code (or primary documentation, even) changes may already have been merged at this point. So it would mean _always_ having a separate issue that was the "true" tracking point on the plan, with the PR just for the code review -- unless, of course, there was no change to UI display. As an alternative (lighter weight and now I've thought, probably preferable), we may choose to say that unless the translations, where needed, are on the PR, the PR won't be merged. I don't feel comfortable with "dumping" the release onto the translators and saying "OK, sort that out". Being part of the development of a feature seems more inclusive. But I can't speak for the translators, of course!

If I understand you correctly, then I disagree.
* As an implementer of a feature with a GUI element, I would be frustrated at a merge having to wait until _all_ the translators had been able to supply their translation. A submitter of a PR from outside the team probably doesn't even know who all the translators are.
* As a translator, I might find it tedious to keep getting asked for individual translations to be done as features trickle in. I would prefer to have all the updates as a batch when preparing for a new release, so I could do them in a single session. And the volume of work required at that time is still not arduous, since it is just changes since the previous release.

> Cycles, rhythm, etc... Not something I'm keen on. KISS should apply. Don't do too much at once. If that means fix immediate bugs plus get the "most ready" feature delivered and nothing else (but including all the docs and translations) every couple of weeks, fine. If we decide that's too often, it's easier to slow down than speed up - just tag a beta and move on.

Yes, I would find target dates too much of a pressure for a volunteer team. I would prefer that we tackled bugs and submitted PRs in priority order (as agreed between us), and when we have either collected enough release-worthy features or fixed a serious bug, we do a code freeze and start the translate/release process.

> Tools... I used both the checklist and Tracking Project. Finding them was the most frustrating part. Maybe have another repo -- "jamulussoftware/pm" -- where the `README.md` has "Useful links for project management" and little else... :)

I mainly used the checklist and Milestones. I'm not yet used to looking at the Projects page, but I must remember to do so.

> I'd rather not have to review _every_ discussion -- I don't have time -- but I do want to see the Issues, to check whether I think they'll have significant impact one way or another. Yes, it may end up back in the discussion getting the issue closed (but at least I'll know that one needs following).

Yes, I'm also still getting used to also checking Discussions, rather than just Issues and PRs!

**Reply by jerogee** [reply]
>  As a translator, I might find it tedious to keep getting asked for individual translations to be done as features trickle in. I would prefer to have all the updates as a batch when preparing for a new release, so I could do them in a single session. And the volume of work required at that time is still not arduous, since it is just changes since the previous release.

I agree, at least that is what I prefer for the Dutch translation. I quite liked the "label freeze" heads up that Volker used to give a few weeks before release: a heads up that from now on no new labels or label changes will occur  so as a translator you can choose to batch translate all changes at once or work on it incrementally until release time. If I were to do this for each feature that trickles in, I would spend at least twice the time on the translation process because of the overhead and having to be on the standby during the full development cycle (I am quite sure not all translators will be able to do that).

**Reply by gilgongo** [reply]
@softins @jerogee 

>  I would be frustrated at a merge having to wait until all the translators had been able to supply their translation

Well yes, but are you implying that if some translators couldn't supply their translation then we would release without it?  

>  As a translator, I might find it tedious to keep getting asked for individual translations to be done as features trickle in

Forgive me, but where or how is that implied by this discussion? I took it to mean exactly the opposite! :-) We "label freeze" to get translation done on the app in one go, and cut a dedicated branch on the website for same reason. 

>Yes, I would find target dates too much of a pressure for a volunteer team

But the minute we make any time-relative statement like "in the next two weeks", that implies a date. So why not just state that date so everyone's clear? It's not as if there's some terrible consequence of missing it by a few days (in fact didn't Volker used to explicitly name a date for the release to give notice to the translators?)

**Reply by softins** [reply]
> @softins @jerogee
> 
> > I would be frustrated at a merge having to wait until all the translators had been able to supply their translation
> 
> Well yes, but are you implying that if some translators couldn't supply their translation then we would release without it?

Well we did that this time with Slovak.

But no, I'm not talking about releases, but merging of individual PRs during the time between releases. I was responding to what I thought was Peter's suggestion that any PR affecting GUI text should include all the translations of that text before it could be merged.

> > As a translator, I might find it tedious to keep getting asked for individual translations to be done as features trickle in
> 
> Forgive me, but where or how is that implied by this discussion?

That's what I thought Peter was suggesting. I may have misunderstood.

> I took it to mean exactly the opposite! :-) We "label freeze" to get translation done on the app in one go, and cut a dedicated branch on the website for same reason.

Yes, that's what we did this time, and what I would prefer for the future too.

> > Yes, I would find target dates too much of a pressure for a volunteer team
> 
> But the minute we make any time-relative statement like "in the next two weeks", that implies a date. So why not just state that date so everyone's clear? It's not as if there's some terrible consequence of missing it by a few days (in fact didn't Volker used to explicitly name a date for the release to give notice to the translators?)

Ah, I was referring to the idea of a schedule of dates, such as "every month" or "every 6 weeks" or whatever. I have no problem with setting an individual target date when we have decided we have good reason to do a release.

**Reply by gilgongo** [reply]
>  Peter's suggestion that any PR affecting GUI text should include all the translations of that text before it could be merged.

Yes, I didn't understand that either...

> I have no problem with setting an individual target date when we have decided we have good reason to do a release.

OK, so do you mean we'd look at the stuff on the backlog and say, "Let's do A, B and C and that feels like three weeks, so next release is %DD%MM"? And then after that's gone out we might say, "Let's do X, Y and Z but those feels like eight weeks"? That  seems like it would be harder to organise and agree than just saying "Next release %DD%MM", and then whatever gets done by then gets in. Obviously, we'd need suitable code freeze and a translation windows every time though. Predictable dates seem obviously easier to deal with for people like translators or anyone not closely involved, is all.

**Reply by pljones** [reply]
> You say it's not something you're keen on, but then describe a two-week cycle :-)

"Every couple of weeks" isn't "a two-week cycle" -- it means "it's ready when it's ready" but we're not trying to do too much.

> Not sure I understand - do you mean creating a PR for each language in advance
> If I understand you correctly, then I disagree.

Tony got it.  I meant as part of the change.  That way, master should always be ready to release.  Why should any language take priority over any other for getting to master?

> Well yes, but are you implying that if some translators couldn't supply their translation then we would release without it?

Exactly.

> But the minute we make any time-relative statement like "in the next two weeks", that implies a date.

So don't.  Say "when this change is complete, it can be released" - and define what "complete" means.

And do one change at a time - to master.  KISS.

Once you're familiar with git, keeping a feature branch up to date to avoid future merge conflicts is _not difficult_.  It shouldn't be used as an excuse to avoid keep changes coherent.

**Reply by gilgongo** [reply]
Ok so "it's ready when it's ready". The down side of that is that it requires continuous monitoring by those not as close to the project as we are. But maybe that's not a big deal.

> Tony got it.

But disagreed? I myself don't know what it would mean in practice so I can't comment.

As to release without a translation (in the app, not sure if you also meant in the website), we'd just show the untranslated strings, is that right? Better than disabling the language selection I guess. 

Be that as it may, I think it's worth agreeing that we are not in fact committed to translation as a dependency of releases. Not sure how to handle that on the web site though.

**Reply by pljones** [reply]
OK, I still think it's worth alerting translators when a change is made that will mean the translation strings need updating.

The same applies with the documentation -- for any change, the existing documentation that applies should be considered and a decision taken as to whether to do it "now" (as part of the application change) or "later" (as part of the release process and requiring additional tracking) - plus the translations then cascading from there.

To me, _not_ keeping it all together makes the whole thing more chaotic and prone to error.  That could be because that's what happens if I try to work that way..! :D

**Reply by gilgongo** [reply]
>  I still think it's worth alerting translators when a change is made that will mean the translation strings need updating

But isn't that "drip feeding" as both @jerogee  and @softins point out? If so, then we have a bit of a major divergence of opinion on this matter! :-)

We could of course abandon the idea of needing to have everything translated in favour of regular quick releases. But compared to the current system we have that does seem a backward step perhaps. Granted, we never really discussed the pros and cons of translation until now, so maybe that time has come!

**Reply by pljones** [reply]
"Alerting" means giving someone a notice that something has happened.  It doesn't tell them they have to do anything about it.  If one translator chooses to wait until just before we want a release out to do everything, that seems to be acceptable here, right?  What if a different translator wants to know what on earth the change was being done for and have more contextual information?  Perhaps they _want_ to see it sooner rather than later, when the developer might not be around to explain.

**Reply by gilgongo** [reply]
OK but when this was discussed before (mainly with @ann0see and @ignotus666 ) it was felt that having a squash-merged branch with all the changes in one place to work though was better - for the website at least. Maybe not for the app though? 

I'd be interested to see what other translators say.

_BTW we need to resolve this issue soon-ish because we were intending to write it up and stick it on the website before the next release._

**Reply by hoffie** [reply]
**Release rhythm:**
I'd be a fan of having a soft target of X weeks after the previous release for two reasons:
- Users know what to expect.
- We avoid "Hrm, should we do a beta now?" discussions. The timing of it felt rather random and rather late for 3.7.0.

I'd be fully OK with a "ok, let's do it earlier because we have some bugfix we should really get out" as well with "ok, let's delay for some more days to get PR X merged".

I'd like to avoid fixed dates ("release every 1st of the month") because that creates too much stress for a volunteer project and is prone to fail if we had delays in the previous cycle.

In theory, we could release after each important PR merge (the recent fader adjustment work by @JohannesBrx or the delay panning by Detlef Hennings would be examples). However, this would also mean that we would have to accept the overhead of translations, our release process (tests, uploads, announcements) and things like Windows Smartscreen acceptance. Therefore, I'd not be a fan of such short release cycles.

**Translation PRs**
I think we should stick to the current approach of "soft freeze", after which no translation-relevant changefs may be touched, followed by a ~two week translation phase. The only thing I think we should change is that we should have per-language issues to simplify tracking (instead of having checklist entries).

While I agree that master should always be in a releaseable state, I think we have to exclude translations from that goal. I see too many downsides and questions:
- Translators would get possibly tens of pings per release.
- How should they add their changes? Would a feature PR have to wait for all translations to settle in? Would translators have to submit PRs against the feature PR's author branch? Wouldn't this also further increase the waiting time for feature PR authors a lot? Or would that be PRs against master? How would we track if/what has been translated?

I don't think we should change anything regarding merging (sqash merge) for the code base because the code translations use specific tooling (linguist) which takes care of showing new stuff to translate (AFAIU).

I do agree that we should accept that some languages may not be translated for a release (as we did for 3.7.0). However, I think we should still aim to cover most of it. If 50% of all translations are still missing after two weeks, we may re-ping and wait some more days. In other words, just be flexibel. Translators are volunteers as well, after all. :)

**Versioning**
While we are at it, should we consider adopting something like [Semver](https://semver.org/)? The last version 3.7.0 was a feature release. The next version is planned to be 3.7.1, although it will be a feature release as well?
I would welcome something like:
- Major version (i.e. Jamulus 4): Possibly incompatible, architectural changes
- Minor version (i.e. Jamulus 3.8): The next feature release (delay panning, fader adjustments)
- Patch level version (i.e. Jamulus 3.7.1): Bugfixes, especially regressions such as newly introduced crashes, late translations.

**Reply by gilgongo** [reply]
> I'd be a fan of having a soft target of X weeks after the previous release
> ...
>  I'd not be a fan of such short release cycles.

OK so if we test this by looking at what's currently [tagged for 3.7.1](https://github.com/orgs/jamulussoftware/projects/2?card_filter_query=milestone%3A%22release+3.7.1%22).  Based on the status of those things - what does that feel like? 6 weeks? 8 (+ 2 for translation)? 

I assume we'd say "in progress" takes precedence over "backlog" BTW if we wanted to de-tag, since I'm assuming everything in the backlog is not being worked on right now. That may not be a safe assumption though since it's mostly me assigning statuses to things. That'll stop soon though once people start collaborating on the tracking, won't it ;-)

And if we test it with @pljones method, what does that look like? Do the majority of things there get de-tagged based on a vote for the most important ones? 

Would be good to get some detail on how things might work in practice.

**Reply by pljones** [reply]
> things like Windows Smartscreen acceptance

That's probably the strongest argument _against_ short release cycles there is...

**Reply by pljones** [reply]
> Versioning

I've no strong views - I've seen too many different weights, flavours and flame wars to worry :). My own views?
* Major -- breaking change: we're stating _something_ will be incompatible / you can't go backwards -- save your configs!
* Minor -- non-breaking change: includes new features (because we do no break compatibility) -- could be one or two things but try not to do too much (as these are likely to be "big" changes)
* Patch -- feature improvements, UI tweaks with no functional change, fixes -- but these need to be sufficiently substantial to make a release worthwhile, so likely to include more than one thing

As you can see, it's all a big vague...

**Reply by pljones** [reply]
> Tracking for 3.7.1

There's a bunch of stuff tagged.  Does that mean "committed to deliver"?  Or did the 3.7.1 tag get added to mean "definitely not 3.7.0"?

So I think a clean up needs to be done first.  For example there are a couple of bugs in Backlog that I don't know whether they've been accepted as "will fix" or "won't fix" yet, so I'm not sure how they can have a release tag at all.

There was a bunch of stuff "nearly ready" but just missed code freeze for 3.7.0, so it would be good to get that completed, merged and closed off, so it's "Done" on the board.

Both of those should let us get a better idea of whether we want to have more change in a 3.7.1 release (or a 3.8.0 release as there's a couple of "new features"?).

I'd still like to keep the _existing_ ideas getting through into release versions as quick as we can - I'm keen to keep the board as clear as possible.  However, the amount of change in the last six months is unprecedented (perhaps it will continue, maybe things will slow down again; who knows?), so I may be being over-optimistic and each release is going to have to be larger than had been the case.

**Reply by gilgongo** [reply]
> There's a bunch of stuff tagged. Does that mean "committed to deliver"?

Exactly :-) I think we all have a slightly different view of what qualifies for a release or a state. Those bugs you refer to - I may have moved to backlog because I saw somebody put a release on them. And if if has a release it can hardly be in triage - so moved to next stage. We can ask the people who tagged them why they did it? Eg https://github.com/jamulussoftware/jamulus/issues/1289 @ann0see 

> There was a bunch of stuff "nearly ready" but just missed code freeze

We could start indicating priority with positions in the "in progress" column (at the risk of more "process"...)

**Reply by pljones** [reply]
I guess I'm saying any issue/PR shouldn't be release tagged _until_ we've committed to getting it into that release.

So we build from an "empty" release X.Y.Z, taking each ticket in turn and deciding "yes/no" as "committed" (meaning "yes, we will _delay release_ for this") to a planned release.  Obviously there's still flexibility to drop / add as situations change.  But the tag shouldn't be "aspirational" - that just causes clutter (more of my fear of untidiness creeping through).

**Reply by gene96817** [reply]
It is really incredible how hard the development team works. From the non-technical user's perspective, a release every quarter (every 90 days) is a bit overwhelming. It gives them new software anxiety too often. This also runs counter to some users' defensive measure of never installing software less than 6 months after a release. 

I see a lot of good bug fixes and new features in the pipeline. I am torn on how to get the improvements out and how to manage a chorus. The challenge is some users will be eager for a feature but that drags the reluctant (and high support effort) user along. In the end, I think we have to choose to minimize user anxiety and support.

**Reply by gilgongo** [reply]
> the tag shouldn't be "aspirational" 

I'd agree with that. There's a tendency to want to do everything for the next release, which needs to be curbed a bit :-) I can feel that road map thing coming on again...

> From the non-technical user's perspective, a release every quarter (every 90 days) is a bit overwhelming.

That's a very good point. Do we have people asking us when the next release will be and what it will contain? Well, er .. no actually.

**Reply by hoffie** [reply]
@pljones 

> There's a bunch of stuff tagged. Does that mean "committed to deliver"? Or did the 3.7.1 tag get added to mean "definitely not 3.7.0"?

I certainly used this as "won't make it into 3.7.0, so it'll be 3.7.1 for now".

> So I think a clean up needs to be done first.

Yes, we should agree on how to use that.

> For example there are a couple of bugs in Backlog that I don't know whether they've been accepted as "will fix" or "won't fix" yet, so I'm not sure how they can have a release tag at all.

I was hoping to be able to use Backlog as the list of things which definitely should be done (at some point). Undecided items should be in Triage. I'm not sure that this is currently the case.

Neither am I sure who should do the moving? There are obvious bugs where a single person may decide this with high confidence that everyone on the team would agree. But who gets to decide whether we'd accept a specific feature? I'm not sure if we should require each main developer to comment on each such issue. On the other hand, this needs at least some visibility. Project board column changes don't trigger any notifications, afaik. So if developer A moves an issue to the backlog (and therefore decides to accept the feature), this information will not be propagated to other developers although it should be.

I think we should have "Lazy consensus" for that. Let a single developre decide, but wait X (5?) days for other developers to chime in and veto if necessary. I'm still not sure how to use Github tools to document/signal that. At worst, it will have to be a comment "Sounds good, I'd like to have it, let's wait for other developers to agree" (which triggers a notification). As long as this is the case, the issue would stay in Triage.

> There was a bunch of stuff "nearly ready" but just missed code freeze for 3.7.0, so it would be good to get that completed, merged and closed off, so it's "Done" on the board.

Yes.

> I guess I'm saying any issue/PR shouldn't be release tagged _until_ we've committed to getting it into that release.
> So we build from an "empty" release X.Y.Z, taking each ticket in turn and deciding "yes/no" as "committed" (meaning "yes, we will _delay release_ for this") to a planned release. Obviously there's still flexibility to drop / add as situations change. But the tag shouldn't be "aspirational" - that just causes clutter (more of my fear of untidiness creeping through).

That's not how I hoped it would work (see above). I'm not totally opposed either, but I do have questions if we go that route. For example: How would we commit to getting it into a release? Would this require synchronuos planning meetings?

What I'd like:
- Be explicit what we want to have in Jamulus at all. Decide about new features as early as possible if to keep the board clean and to avoid unnecessary work for contributors. We need to sort out a process for that as described above ("Lazy conensus").
- Tag anything which is close to ready and/or which has enough activity for the next feature release. If the code freeze is nearing and things are still not done (but there are enough valuable PRs merged), move them to the next feature release milestone.

This approach has the downside that the Milestone completion indicator will only be useful after the code freeze.
This approach has the upside that it is efficient in regards to our and our contributor's energy or flow:
- Inactive or slow PRs will move to later releases (we can't control how much time contributors invest anyway, except for deciding to finish PRs ourselves, which should just be a last resort).
- Active PRs stay active. Needlessly putting them on hold makes it harder for us as we will have to invest energy to activate them again. It will also be a worse experience for contributors who are hoping to get their features in.

This is a different approach to what feature development at work may look like. However, at work there is time to allocate and developers ("contributors") are "forced" to work on the decided features whereas we are dealing with volunteers here. Nothing prevents us to try to get certain PRs in with higher priority. This could be done by individual main developers by reviewing early, making concrete suggestions for improvements and explicitly pinging another developer for the second review.

@gene96817 Thanks for your input as well! So we would have to take into account that too many releases may be a problem for people as well. I understand that when looking at my choir. At the same time, I'm eagerly waiting for the fader adjustment feature to become available in an official release. So it all depends on the perspective. Users may still choose to skip a release and just go for the next one.

**Reply by gene96817** [reply]
> Do we have people asking us when the next release will be and what it will contain? Well, er .. no actually.

I work to avoid discussion of the next release with my users. I want my users to use what we have. No reason for perfect to be the enemy of the good. 

> At the same time, I'm eagerly waiting for the fader adjustment feature to become available in an official release.

Ah yes, so many improvements. It is hard to know they exist and still wait patiently.

**Reply by hoffie** [reply]
Another argument regarding translation handling: If we wanted translators to be able to submit translations without being in a "please translate" phase, we would have to tell them to run `lupdate` themselves or do that in CI.

**Reply by pljones** [reply]
> do that in CI.

I'm hoping this will be done in the 3.7.1 cycle.

**Comment by ignotus666** [comment]
> OK but when this was discussed before (mainly with @ann0see and @ignotus666 ) it was felt that having a squash-merged branch with all the changes in one place to work though was better - for the website at least. Maybe not for the app though?

Yes, no need for squash/merge for the app as Qt Linguist shows us what's translated/untranslated.

It seems like we're bundling app and website translations together here when talking about releases, but I'd argue they're completely different beasts. App translations are relatively trivial, don't amount to more than a few new strings with each new release (unless it's lagged behind) and there's no need to trawl through docs to find the changes as Qt Linguist does it for you. I think an alert a couple of weeks before the release works fine, as @corrados used to do. I wouldn't hold back PRs or releases because of app translations - if they're done in time, great, if not they can get merged into master if/as they arrive. I don't think the odd untranslated menu entry should be a showstopper for anyone.

The website, however... most translations are incomplete/lagging behind with outdated information, and my hunch is that this is only going to get worse. The reason is that with every new release, the list of edits just piles up and makes checking all of them an impossible task as old squash/merges fade into the ether. It's not so much the translation work itself (which is a fairly big task in its own right) as the mind-numbingly tedious work of trawling through docs and scanning them for differences between the original in English and the translation, which is what you have to resort to when your language lags by one release or more. I'll raise my hand and admit I should have foreseen this was going to happen and proposed a system that would have made the workflow more similar to that of Qt Linguist, by using a CAT tool like [OmegaT](https://omegat.org/) (what I use) or [GitLocalize](https://gitlocalize.com/) (done some tentative testing and may be less intimidating than OmegaT). 

However, at this point, migrating to either of these translation memory-based systems would involve a considerable amount of work, as existing translations would first have to be aligned (this means sentence-level alignment of existing translated segments, not actually translating everything missing to date - OmegaT can do this automatically with some manual input) to produce a functional translation memory (.tmx) file, and translators would have to familiarise themselves with the tool chosen... In all scenarios (including the current), it's a massive amount of work, which begs the question - have we bitten off more than we can chew? Stagnant/outdated translations will inevitably lead to problems when people start following instructions that may no longer apply.

Sorry for veering a bit OT but I think it's kind of tangentially related to the release process.

Tl;dr:
The system for app translations (I think) has worked fine to date with the 2-week notice. The website translation workflow might need a review/overhaul.

It'd be interesting to know what other @jamulussoftware/translators think.

**Reply by gilgongo** [reply]
>  it's a massive amount of work, which begs the question - have we bitten off more than we can chew?

It's certainly observable that very few projects, FOSS or not, have fully-localised websites. Maybe there's a reason for that? 

One thing to consider is that I **think** the website (English) content **should** settle down over time once we iron out all the current deficiencies (eg the Linux install page needs gutting, the Getting Started and Oboarding pages confusingly overlap, we don't have a release process, team page, and the contributing pages are in a mess and are being re-written now). 

So maybe let's pull the localisation until all that's in the clear in about a year or so?

**Reply by pljones** [reply]
Yes, it does seem counter-intuitive to be spending effort translating site content where it's known that it's out of date anyway.  Better to put effort (in any language) into bringing the content up to date (and, if needed, porting to English so the "base" language is there -- I've no idea how the tools work).

**Reply by gilgongo** [reply]
OK so let's wait for others to have their say in this.

But if we are generally in agreement I think we should remove the language switching on the website until we can get the English to a point where only small incremental changes are happening. 

Maybe give it another week or so from today for comments and counter-proposals.

**Reply by jerogee** [reply]
So far I have only been involved in app translation, but as for website translations I agree with @gilgongo: there is a lot to say for first correcting, stabilizing, and maturing the site content in English to the point that no major revisions are needed anymore. And to only then turn to (or revisit) localisation, and for that only to go for languages that can be properly resourced (enough translators), with the principe: rather few done really well than many half-baked, outdated, and erroneous translations (no matter the noble effort and how well intended) that only confuse Jamulus users.

**Reply by jujudusud** [reply]
Be carrefull, please consider that some documentations (thinking about linuxmao.org) are already pointing to the french translation pages of the website. 
From my point of view, it's not a good thing to let translated pages down for a year...

**Reply by ann0see** [reply]
Same for German. I've only translated a few pages (only the ones I consider most important) and I'd say that we could handle it like that.

So probably only

1. Homepage
2. Getting Started
3. Windows/Mac/Linux installation pages need to be updated.

Everything else can and stay untranslated unless someone wants to do it.

**Reply by gilgongo** [reply]
@ann0see  OK, but we don't have stable English versions of 2 and 3 yet, so we'd be asking people to update them in the knowledge that we will be re-writing them again (2 as a result of @mulyaj's work https://github.com/jamulussoftware/jamuluswebsite/issues/349 and https://github.com/jamulussoftware/jamuluswebsite/issues/322, and 3 because binary installs are what we want from now on). 

It's going to be a lot less problematic to just remove localised pages now, then introduce them once they are in a maintainable state. Incoming links can be re-directed if we know about them. 

I can do a PR for 3 soon, but 2 is going to take a bit more time.

**Reply by ann0see** [reply]
Probably true. But 2 and 3 are important (at least for me) so I'd take the work and translate them although they might be updated soon.

**Reply by jujudusud** [reply]
I will be pleased to help translate the french part.
I think it is not a problem to translate only one part when it is in stable branch and wait to translate the other part when it comes to stable branch.

**Reply by ann0see** [reply]
Yes. That would be great. Tagging @trebmuh and @gegeweb here (they've been the translators until now).

**Reply by gilgongo** [reply]
So if people are willing to translate what's there now on the understanding that 2 and 3 will need completely re-translating not too long after, then that's fine. 

1. Homepage
2. Getting Started
3. Windows/Mac/Linux installation pages

Can we agree to pause the rest of the site now until things settle after that?

**Reply by trebmuh** [reply]
> Yes. That would be great. Tagging @trebmuh and @gegeweb here (they've been the translators until now).

I'll probably have some time on the coming days for the website translation work.
I already know pretty well @jujudusud and will get him on board when starting translating.

**Reply by gilgongo** [reply]
@trebmuh Just so we all understand: the proposal is to **only** translate (and keep up to date from now on if possible) the 3 sections @ann0see lists, is that right? 

BTW I take it 3 is the client installation page, not the server.

**Reply by ann0see** [reply]
Yes. 3 is the client page.

**Reply by ignotus666** [reply]
In the case of the ES translation, it's all up to date - it'd be a shame to pull it offline that being the case. I'm fairly confident I can keep it that way - particularly if those 3 documents are the ones that are going to see the most changes and the rest will stay more or less the same.

**Reply by gilgongo** [reply]
OK so the only outstanding question is how we selectively remove the language selector from the pages that we're pausing?

**Reply by gilgongo** [reply]
@ignotus666 You ain't seen my re-write... :-) But OK if you're OK with that.

**Reply by ignotus666** [reply]
I reserve the right to eat my words ;)
